File: 207-course-schedule/course-schedule.java
class Solution {
    Set<Integer> current = new HashSet<>();
    Set<Integer> visited = new HashSet<>();
    boolean cycle = false;
    Map<Integer, List<Integer>> map = new HashMap<>();
    void helper(int start){
        if(!map.containsKey(start) || visited.contains(start)){
            return;
        }
        if(current.contains(start)){
            cycle = true;
            return;
        }
        current.add(start);
        List<Integer> pre = map.get(start);
        for(int j = 0; j < pre.size(); j++){
            helper(pre.get(j));
        }
        current.remove(start);
        visited.add(start);
    }
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int i = 0;
        while(i  < prerequisites.length){
            if(map.containsKey(prerequisites[i][0])){
                map.get(prerequisites[i][0]).add(prerequisites[i][1]);
            }
            else{
                map.put(prerequisites[i][0], new ArrayList<>());
                map.get(prerequisites[i][0]).add(prerequisites[i][1]);
            }
            i++;
        }
        for(Map.Entry<Integer, List<Integer>> curr : map.entrySet()){
            if(!visited.contains(curr.getKey())){
                helper(curr.getKey());
            }
            if(cycle){
                return false;
            }
        }
        return true;
    }
}

File: 207-course-schedule/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        hashmap = {}
        visited = set()
        completed = set()
        for item in prerequisites:
            key = item[0]
            val = item[1]
            if key in hashmap:
                hashmap[key].append(val)
            else:
                hashmap[key] = []
                hashmap[key].append(val)
        result = True
        def helper(curr):
            nonlocal result
            if (curr not in hashmap) or (curr in completed):
                return
            if curr in visited:
                result = False
                return
            visited.add(curr)
            for i in hashmap[curr]:
                helper(i)
            visited.remove(curr)
            completed.add(curr)
        for item in hashmap:
            helper(item)
            if not result:
                return False
        return result
        